docs/guide/front-end-helpers.md

```diff
@@ -0,0 +1,319 @@
+// }
+```
+
+### `attachmentUrl(fileObj, options)`
+
+Get the file URL for an Apostrophe attachment object. Optionally pass an options object to get a specific version of the file. File (non-image) attachment objects include a single `_url` property already, so this is primarily used for retrieving specific versions of an image.
+
+| Argument | What is it? |
+| -------- | ----------- |
+| `fileObj` | an attachment object, such as from an attachment field value or using a template helper method to parse an [image widget](/guide/core-widgets.md#image-widget) value |
+| `options` | image options to apply for the resulting URL |
+
+| Options | What is it? |
+| -------- | ----------- |
+| `crop` | an object of image cropping coordinates with `left`, `top` and `width` properties (usually populated from the cropping UI and included on the image object) |
+| `size` | an image size name to retrieve, such as one of the [standard image file variations](/guide/core-widgets.md#specifying-the-fallback-size) |
+
+```javascript
+// Getting an image attachment object after stashing the stringified object
+// on a `data-image` attribute in the template
+const imageObj = document.querySelector('[data-thumbnail]').dataset.image;
+
+const smallImage = apos.util.attachmentUrl(imageObj, {
+  size: 'one-third'
+});
+```
+
+### `closest(el, selector)`
+
+Returns the closest ancestor element that matches the selector. The element itself is considered the closest possible match. Supports browsers without the matching native method.
+
+| Argument | What is it? |
+| -------- | ----------- |
+| `el` | a DOM element |
+| `selector` | a string to use as a CSS selector to match |
+
+```javascript
+const menuToggle = document.querySelector('[data-toggle]');
+
+const menu = apos.util.closest(myElement, '[data-menu]');
+```
+
+### `emit(el, name, data)`
+
+Emit a custom browser event on a DOM element. Optionally include a `data` object to include on the event. For event listeners, use the standard browser [`addEventListener` method](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener).
+
+| Argument | What is it? |
+| -------- | ----------- |
+| `el` | a DOM element |
+| `name` | a string to use as the event name |
+| `data` | (optional) an object containing data to include on the event object |
+
+```javascript
+const myElement = document.querySelector('[data-my-element]');
+
+apos.util.emit(myElement, 'openChat', {
+  url: window.location
+});
+```
+
+### `getCookie(name)`
+
+Get the value of a browser cookie.
+
+| Argument | What is it? |
+| -------- | ----------- |
+| `name` | The name of a browser cookie |
+
+```javascript
+apos.util.getCookie('cookiename');
+```
+
+### `onReadyAndRefresh(fn)`
+
+Runs the function passed in when Apostrophe refreshes page content during editing. When logged out it will run the function on initial page load. This is not necessary in [widget players](/guide/custom-widgets.md#client-side-javascript-for-widgets).
```

Comment:
Runs the function passed in when the page loads, and also when Apostrophe refreshes page content during editing.

Reply:
See also widget players.